source/extensions/quic_listeners/quiche/udp_gso_batch_writer.cc

```diff
@@ -0,0 +1,129 @@
+#include "extensions/quic_listeners/quiche/udp_gso_batch_writer.h"
+
+#include "common/network/io_socket_error_impl.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+
+namespace Envoy {
+namespace Quic {
+namespace {
+Api::IoCallUint64Result convertQuicWriteResult(quic::WriteResult quic_result) {
+  if (quic_result.status == quic::WRITE_STATUS_OK) {
+    // Write Successful
+    if (quic_result.bytes_written == 0) {
+      ENVOY_LOG_MISC(trace, "sendmsg successful, message buffered to send");
+    } else {
+      ENVOY_LOG_MISC(trace, "sendmsg successful, flushed bytes {}", quic_result.bytes_written);
+    }
+    // Return bytes_written as rc & nullptr as error on success
+    return Api::IoCallUint64Result(
+        /*rc=*/quic_result.bytes_written,
+        /*err=*/Api::IoErrorPtr(nullptr, Network::IoSocketError::deleteIoError));
+  }
+
+  if (quic_result.status == quic::WRITE_STATUS_BLOCKED_DATA_BUFFERED) {
+    // Data was buffered, no need to return error
+    ENVOY_LOG_MISC(trace, "sendmsg blocked, message buffered to send");
+    return Api::IoCallUint64Result(
+        /*rc=*/0,
+        /*err=*/Api::IoErrorPtr(nullptr, Network::IoSocketError::deleteIoError));
+  }
+
+  if (quic_result.status == quic::WRITE_STATUS_BLOCKED) {
+    ENVOY_LOG_MISC(trace, "sendmsg blocked, message not buffered");
+    return Api::IoCallUint64Result(
+        /*rc=*/0,
+        /*err=*/Api::IoErrorPtr(new Network::IoSocketError(quic_result.error_code),
+                                Network::IoSocketError::deleteIoError));
+  }
+
+  // Write Failed, return {0 and error_code}
+  ENVOY_LOG_MISC(trace, "sendmsg failed with error code {}",
+                 static_cast<int>(quic_result.error_code));
+  return Api::IoCallUint64Result(
+      /*rc=*/0,
+      /*err=*/Api::IoErrorPtr(new Network::IoSocketError(quic_result.error_code),
+                              Network::IoSocketError::deleteIoError));
+}
+
+} // namespace
+
+// Initialize QuicGsoBatchWriter, set io_handle_ and stats_
+UdpGsoBatchWriter::UdpGsoBatchWriter(Network::IoHandle& io_handle, Stats::Scope& scope)
+    : quic::QuicGsoBatchWriter(std::make_unique<quic::QuicBatchWriterBuffer>(), io_handle.fd()),
+      io_handle_(io_handle), stats_(generateStats(scope)) {}
+
+// Do Nothing in the Destructor For now
+UdpGsoBatchWriter::~UdpGsoBatchWriter() = default;
+
+Api::IoCallUint64Result
+UdpGsoBatchWriter::writePacket(const Buffer::Instance& buffer, const Network::Address::Ip* local_ip,
+                               const Network::Address::Instance& peer_address) {
+  // Convert received parameters to relevant forms
+  quic::QuicSocketAddress peer_addr = envoyAddressInstanceToQuicSocketAddress(peer_address);
+  quic::QuicSocketAddress self_addr = envoyAddressIpToQuicSocketAddress(local_ip);
```

Comment:
local_ip might be nullptr which means kernel decides packet source address. Did you handle this case in envoyAddressIpToQuicSocketAddress()?

Reply:
Returning empty `QuicSocketAddress()` when a `nullptr` is sent to `envoyAddressIpToQuicSocketAddress()`